### PR TITLE
[Feature] Rector Wrapper

### PR DESCRIPTION
**Pull request by 50bhan**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Fixed tickets | #200 

Following V2 road map, this PR will add Rector (https://github.com/rectorphp/rector) wrapper to the project. There are couple of tasks which need to be done to complete this feature:

- [x] Add wrapper
- [x] Increase stability
- [x] Add some default Rectors
- [ ] Add refactor command
- [ ] Add more tests
- [ ] Update doc

P.S: The PR is functional (especially with added Rector and its type), but there are some problem with some rectors at this moment. So I should fix them.

### CODE REVIEW

**Comment by 50bhan** [comment]
@olivernybroe @Jibbarth I tested out the code with all Rectors (literally all 477 of them :smile:) and it seems more stable now. So, I would appreciate if you guys also check it out.

P.S. I made a gist with all Rectors in `functions.php` metric file for testing purposes which you can use: https://gist.github.com/50bhan/86201591fc7a25697c1715473fb0373a

File: src/Domain/Metrics/Code/Functions.php [hunk 1]
```diff
@@ -50,6 +51,7 @@ public function getInsights(): array
             DeprecatedFunctionsSniff::class,
             NullableTypeDeclarationSniff::class,
             StaticClosureSniff::class,
+            ClosureToArrowFunctionRector::class,
```
Comment (Jibbarth) [inline src/Domain/Metrics/Code/Functions.php:54]: It's the only Rector you want to add ? Is there is other useful Rectors that we should add in default configuration ?

Comment (50bhan) [reply src/Domain/Metrics/Code/Functions.php:54]: @Jibbarth Thanks a lot :heart: Yes you should test it out. At this point I really need to make sure it is fully functional.

**Comment by 50bhan** [comment]
@Jibbarth Thanks a lot heart Yes you should test it out. At this point I really need to make sure it is fully functional.

File: src/Domain/Metrics/Code/Functions.php (hunk 1 above)
Comment (50bhan) [reply src/Domain/Metrics/Code/Functions.php:54]: No it's just for demonstration and test. I will prepare a list of Rectors and then we can consider to add them to default config.

**Comment by olivernybroe** [comment]
Awesome work!

Hmm, would it make sense to have like a test that actually tests if it is able to run with all of the rectors?

File: src/Domain/FileProcessors/RectorFileProcessor.php [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                $rector->addErrorDetails($filePath, $e->getMessage());
+
+                continue;
+            }
+
+            $diff = $this->calculateDiff($splFileInfo->getContents(), $newContent);
+
+            if ($diff !== '') {
+                $rector->addDetails($filePath, $diff);
+            }
```
[the remark this answers is not in the document]
Comment (50bhan) [reply src/Domain/FileProcessors/RectorFileProcessor.php:95]: We need to add details only for changed files. Am I right?  I think it's a different matter and I didn't implemented fix (refactor) part yet.

**Comment by 50bhan** [comment]
@olivernybroe Thank you :slightly_smiling_face: Yes we can do that. But I wanted to make sure that the code is functional in real use, so I tested it manually and prepare a gist of all Rectors for testing. If you feel it is necessary, I will add a test with this criteria.

**Comment by roman-eonx** [comment]
@50bhan are you planning to use the latest Rector version (0.11 at the moment)?

**Comment by nunomaduro** [comment]
Allow me to say, that I don't think is a good idea to have this feature at this time.

I think we should focus the time of the maintainers, on improving the existing "features", making Laravel rules more idiomatic for Laravel users, re-working the documentation, etc.

**Comment by 50bhan** [comment]
@roman-eonx As @nunomaduro said, we're not gonna add this feature at the moment 🙂 But if we plan to do this, we'll use the latest version of rector and this PR will be updated based on it.

**Comment by JustSteveKing** [comment]
Closing as a stale PR

**Pull request closed without merging**